### PR TITLE
Avoid filters during construct

### LINKED ISSUE #13496: Infinite loop on Subscriptions extension when trying to add new sub
## EXPLANATION OF THE ISSUE
When you try to add a new subscription, an infinite loop happens.

Setup: Latest master, WP 4.7.3, [Prospress subscriptions](https://github.com/prospress/woocommerce-subscriptions).

Investigation: After reverting https://github.com/woocommerce/woocommerce/pull/13444 locally, this doesn't happen.

## STEPS TO REPRODUCE THE ISSUE
1. Activate plugin WC using latest master
2. Activate extension Subscriptions from Prospress (using their `wc_27_compatibility` branch)
3. Navigate to WooCommerce > Subscriptions
4. Click on `Add Subscription`
Expected: Form shows properly
Actual:
screen shot 2017-03-07 at 15 11 20 pm

### PR DESCRIPTION
**Pull request by mikejolley**
This is an alternative fix for #13496 which avoids filters during construct.

The date paid getter instead returns the correct value in view context, and handles the data when saved.

The status filter is also moved to the getter. It was already handled during save.

@bor0 @thenbrent Can I get your 👁 on this?

Closes #13496

### CODE REVIEW

**Comment by claudiulodro** [comment]
```Parse error: syntax error, unexpected '}' in /home/travis/build/woocommerce/woocommerce/includes/class-wc-order.php on line 736```
You forgot a semicolon. :)

**Comment by mikejolley** [comment]
@claudiulodro I had fixed that and didn't push :D My network went down.

**Comment by bor0** [comment]
LGTM but I will test it in a few.

**Comment by bor0** [comment]
Tests good 👍

**Review by bor0 — APPROVED** [review]

**Comment by thenbrent** [comment]
@mikejolley this looks like a much better solution than #13497. 👍

File: includes/data-stores/class-wc-order-data-store-cpt.php [hunk 1]
```diff
@@ -131,7 +131,14 @@ protected function read_order_data( &$order, $post_object ) {
 	 * @param WC_Order $order
 	 */
 	public function update( &$order ) {
+		// Before updating, ensure date paid is set if missing.
+		if ( ! $order->get_date_paid( 'edit' ) && version_compare( $order->get_version( 'edit' ), '2.7', '<' ) && $order->has_status( apply_filters( 'woocommerce_payment_complete_order_status', $order->needs_processing() ? 'processing' : 'completed', $order->get_id() ) ) ) {
```
Comment (thenbrent) [inline includes/data-stores/class-wc-order-data-store-cpt.php]: `WC_Order::needs_processing()` is a protected method.

**Pull request merged**